### DISCUSSION TITLE
Why spdlog-rs is so fast?

### DISCUSSION THREAD

**Discussion by nneesshh** [SpriteOvO/spdlog-rs#26] [General]
It is wonderful work, but about the benchmark:
"spdlog-rs (0.3.0)
Sync

[info] **********************************************************************
[info] Multi threaded: 1 threads, 250000 messages
[info] **********************************************************************
[info] basic_mt                       Elapsed: 0.09 secs          2881439/sec
[info] rotating_mt                    Elapsed: 0.09 secs          2810747/sec
[info] daily_mt                       Elapsed: 0.09 secs          2829965/sec
[info] level-off                      Elapsed: 0.00 secs        407365162/sec
...

C++ spdlog (1.11.0)
Sync

[info] **************************************************************
[info] Multi threaded: 1 threads, 250,000 messages
[info] **************************************************************
[info] basic_mt                       Elapsed: 0.13 secs        1,906,872/sec
[info] rotating_mt                    Elapsed: 0.13 secs        1,866,094/sec
[info] daily_mt                       Elapsed: 0.13 secs        1,981,681/sec
[info] level-off                      Elapsed: 0.00 secs      147,466,525/sec
...
"

It is amazing that spdlog-rs is much faster than spdlog-c++, how cound it be?

**Comment by SpriteOvO** [comment]
I'm not so quite remember of the details, but `spdlog-rs` is not "much" faster than C++ `spdlog` in some cases.

As you can see, the `flexible-string` version of `spdlog-rs` and `fmt` version of C++ `spdlog` from the benchmark results:

`spdlog-rs` (with `flexible-string` feature):

```
[info] **********************************************************************
[info] Multi threaded: 1 threads, 250000 messages
[info] **********************************************************************
[info] basic_mt                       Elapsed: 0.08 secs          2969371/sec
[info] rotating_mt                    Elapsed: 0.08 secs          2952681/sec
[info] daily_mt                       Elapsed: 0.08 secs          2992041/sec
[info] level-off                      Elapsed: 0.00 secs        514721021/sec
[info] **********************************************************************
[info] Multi threaded: 4 threads, 250000 messages
[info] **********************************************************************
[info] basic_mt                       Elapsed: 0.07 secs          3496977/sec
[info] rotating_mt                    Elapsed: 0.07 secs          3399459/sec
[info] daily_mt                       Elapsed: 0.07 secs          3364660/sec
[info] level-off                      Elapsed: 0.00 secs        632751202/sec
```

C++ `spdlog` (with `fmt` library):

```
[info] **************************************************************
[info] Multi threaded: 1 threads, 250,000 messages
[info] **************************************************************
[info] basic_mt                       Elapsed: 0.07 secs        3,430,498/sec
[info] rotating_mt                    Elapsed: 0.07 secs        3,481,225/sec
[info] daily_mt                       Elapsed: 0.07 secs        3,563,268/sec
[info] level-off                      Elapsed: 0.00 secs      138,757,839/sec
[info] **************************************************************
[info] Multi threaded: 4 threads, 250,000 messages
[info] **************************************************************
[info] basic_mt                       Elapsed: 0.11 secs        2,214,753/sec
[info] rotating_mt                    Elapsed: 0.12 secs        2,059,991/sec
[info] daily_mt                       Elapsed: 0.12 secs        2,058,245/sec
[info] level-off                      Elapsed: 0.00 secs      139,938,427/sec
```

Both of them are much faster than the corresponding standard version, the reasons are the same. `fmt` library provides an on-stack growable buffer type [`basic_memory_buffer`](https://fmt.dev/latest/api.html#_CPPv4I0_6size_t0EN3fmt19basic_memory_bufferE) for formatting string storage ([used here](https://github.com/gabime/spdlog/blob/31cf79a70dfa61a4b2b7509eecca6a3b269bfc4a/include/spdlog/common.h#LL168C36-L168C36)), `flexible-string` provides a [`FlexibleString`](https://docs.rs/flexible-string/0.1.0/flexible_string/struct.FlexibleString.html) as well ([used here](https://github.com/SpriteOvO/spdlog-rs/blob/f076b7d08fc0fe067389b27461c8bbade6144d3f/spdlog/src/string_buf.rs#L32)).

For the on-stack string storage version, C++ `spdlog` is a little faster than `spdlog-rs` for `_mt` sink with a single thread. For the performance degradation of C++ `spdlog` in 4 threads, I guess it's caused by some concurrent data structure implementation, but it's just a guess, I didn't verify it.

For your question, why the standard version of `spdlog-rs` is much faster than C++ `spdlog`, if I remember correctly, the main reason is `spdlog-rs` do a heap pre-reservation ([here](https://github.com/SpriteOvO/spdlog-rs/blob/f076b7d08fc0fe067389b27461c8bbade6144d3f/spdlog/src/formatter/full_formatter.rs#L60)) for formatting string storage, but C++ `spdlog` do not. A reservation can significantly improve performance by reducing fragmented heap allocation when formatting records.
